### PR TITLE
DSV4 NNX support

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

This PR adds DeepSeek V4 (`DEEPSEEK4`) decoder block support and HyperHead routing integration to `NNXDecoder`.

### Context & Implementation Details
- Instantiates `DeepSeek4HyperHead` on `NNXDecoder` when `mhc_expansion_rate > 1` and `decoder_block == DecoderBlockType.DEEPSEEK4`.
- Routes the post-transformer hidden state reduction in `NNXDecoder` through `self.hc_head(y)` for `DEEPSEEK4` instead of the default `mhc_reduce(y)`.
- Added `DecoderBlockType.DEEPSEEK4` to the list of supported decoder block types in `NNXDecoder`.

# Tests

- Added unit test `test_nnx_decoder_hyperhead_integration` in `tests/unit/deepseek_v4_vs_reference_test.py` to verify `NNXDecoder` initializes `hc_head` as `DeepSeek4HyperHead`.
- Ran `pytest tests/unit/deepseek_v4_vs_reference_test.py`.
- [comparison of training loss curve over 10 steps vs linen reference](https://b.corp.google.com/issues/535324012#comment6)

# Checklist

- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/nnx_decoders.py
Comment (dipakg-lang) [inline src/maxtext/layers/nnx_decoders.py:462]: Shouldn't we changes inside this method for properly constructing the 43 layer for Deepseek V4 with first three containing hash routing? this current implementation would cause to actually create 86 wrong layers

Comment (dipakg-lang) [reply src/maxtext/layers/nnx_decoders.py:462]: You can also refer to the Linen implemention in decoders.py file for guidance that does correctly handle this part.

**Review by dipakg-lang — COMMENTED** [review]
Thanks for making the changes, can you also test the the flash model on a TPU cluster and add the loss curve to the commit? the changes looks good to me otherwise and I can approve them with the loss curve details here.

File: src/maxtext/layers/nnx_decoders.py
Comment (parambole) [reply src/maxtext/layers/nnx_decoders.py:462]: Hey @dipakg-lang thanks for calling this out. @snehalv2002 Can you make those changes as part of this PR ?

Thanks!

Comment (snehalv2002) [reply src/maxtext/layers/nnx_decoders.py:462]: done

File: src/maxtext/layers/nnx_decoders.py [hunk 1]
```diff
@@ -532,13 +540,39 @@ def _init_scanned_layers(self, decoder_block_classes, rngs, mesh):
...
       self._init_scanned_gemma3(decoder_block_classes, rngs, mesh)
     elif self.is_gemma4:
       self._init_scanned_gemma4(decoder_block_classes, rngs, mesh)
     else:
       self._init_scanned_generic(decoder_block_classes, rngs)
 
+  def _init_scanned_deepseek4(self, rngs):
+    """Initializes DeepSeek V4 scanned layers: unrolls first_num_hash_layers prefix layers and scans remaining full blocks."""
+    config = self.config
+    self.layers = nnx.List([])
```
Comment (parambole) [inline src/maxtext/layers/nnx_decoders.py]: nit: 

Either remove self.layers = nnx.List([]) from _init_scanned_deepseek4 (matching how _init_scanned_deepseek defines self.dense_layers and self.moe_layers without self.layers), or document why an empty self.layers attribute is intentionally left on the module.

File: tests/unit/deepseek_v4_vs_reference_test.py [hunk 2]
```diff
@@ -1512,6 +1512,18 @@ def test_hyper_head_parity(self):
     print(f"HYPER HEAD PARITY - MAX ABS DIFF: {max_diff:.6e}, MEAN ABS DIFF: {mean_diff:.6e}")
     np.testing.assert_allclose(mt_out, pt_out, rtol=5e-5, atol=5e-5)
 
+  def test_nnx_decoder_hyperhead_integration(self):
```
Comment (parambole) [inline tests/unit/deepseek_v4_vs_reference_test.py:1515]: nit:

Expand test_nnx_decoder_hyperhead_integration to run a forward pass through mt_decoder(...) with a dummy input and assert that mt_decoder.hc_head.__call__ was invoked (e.g. via unittest.mock.patch.object(DeepSeek4HyperHead, '__call__', return_value=...) or asserting on the output shape/values).

**Review by parambole — APPROVED** [review]
LGTM. 

I have left a few nits please PTAL.

**Review by entrpn — APPROVED** [review]

**Review by dipakg-lang — APPROVED** [review]

File: src/maxtext/layers/nnx_decoders.py [hunk 3]
```diff
@@ -432,10 +432,18 @@ def __init__(
 
     self.scanned_layers = None
     self.is_deepseek = self.config.decoder_block == DecoderBlockType.DEEPSEEK
+    self.is_deepseek4 = self.config.decoder_block == DecoderBlockType.DEEPSEEK4
     self.is_gemma3 = self.config.decoder_block == DecoderBlockType.GEMMA3
     self.is_gemma4 = self.config.decoder_block == DecoderBlockType.GEMMA4
     self.is_gemma4_small = self.config.decoder_block == DecoderBlockType.GEMMA4_SMALL
 
+    if config.mhc_expansion_rate > 1 and config.decoder_block == DecoderBlockType.DEEPSEEK4:
```
Comment (RissyRan) [inline src/maxtext/layers/nnx_decoders.py:440]: Why do we include `config.mhc_expansion_rate > 1` here? I think it's already included: https://github.com/AI-Hypercomputer/maxtext/blob/8fcc9e133da3ef6e6b46e59121bb318fb6f1db36/src/maxtext/configs/models/deepseek4-284b.yml#L33

**Review by RissyRan — APPROVED** [review]
LGTM at high level. Do we have compile test for v4 yet?

File: src/maxtext/layers/nnx_decoders.py (hunk 3 above)
Comment (snehalv2002) [reply src/maxtext/layers/nnx_decoders.py:440]: I thought about it, can't hurt to leave it in, in case someone wants to override model parameters and run with expansion rate = 1
